### DISCUSSION TITLE
Problem with IF, IFELSE Statements in custom build firmware

### DISCUSSION THREAD

**Discussion by jackpotcityco** [arendst/Tasmota#19771] [General]
Hello,

I think I have managed to first successfully `build` a custom firmware that I also have managed to successfully `upload` to my board: `ESP8266EX` aswell.
I had a successfull upload using the `esptool`. 

Not showing exactly every detail (perheps its not needed for this question). I will show perheps the essential parts if its possible to see the problem anyway. **_(The questions are 4 and 5 further down)_**

1. I have cloned this into the folder: `C:\Tasmota`  ` (git clone https://github.com/arendst/Tasmota.git)`
2. In the `C:\Tasmota\platformio.ini` I have updated the `[common]` section like this where I for example added: `board = nodemcuv2`
```
[common]
platform = ${core.platform}
platform_packages = ${core.platform_packages}
framework = arduino
board = nodemcuv2  ; Use a default board name
board_build.flash_mode = dout
board_build.flash_freq = 40m
board_build.flash_size = 4096KB  ; Updated to match your board
board_build.filesystem = littlefs
custom_unpack_dir = unpacked_littlefs
build_unflags = ${core.build_unflags}
build_flags = ${core.build_flags}
monitor_speed = 115200
monitor_echo = yes
upload_resetmethod = nodemcu
extra_scripts = ${esp_defaults.extra_scripts}
lib_archive = no
lib_ldf_mode = chain
lib_compat_mode = strict
shared_libdeps_dir = lib
lib_extra_dirs =
                              lib/lib_basic
                              lib/lib_i2c
                              lib/lib_display
                              lib/lib_ssl
                              lib/lib_audio
                              lib/lib_rf
                              lib/lib_div
```

3. I created the file: `"C:\Tasmota\tasmota\user_config_override.h"` which has below information to support the `VL53L1X` distance ToF sensor and `IF_STATEMENTS`
```
#ifndef _USER_CONFIG_OVERRIDE_H_
#define _USER_CONFIG_OVERRIDE_H_

// Enable support for VL53L1X
#ifndef USE_VL53L1X
#define USE_VL53L1X
#endif

// Enable support for expressions
#ifndef USE_EXPRESSION
#define USE_EXPRESSION
#endif

// Enable support for IF statements
#ifndef SUPPORT_IF_STATEMENT
#define SUPPORT_IF_STATEMENT
#endif
#endif  // _USER_CONFIG_OVERRIDE_H_
```

4. I now try to use an `IF STATMENT` like below: But it shows: `RESULT = {"Var1":"DistanceA,21.1"}` even that the Distance is not <=15.
`Rule1 ON VL53L1X#Distance DO IF (VL53L1X#Distance<=15) var1 DistanceA,%value% ENDIF ENDON`
![image](https://github.com/arendst/Tasmota/assets/47510843/4c9a1a5a-f3d9-44fc-b144-1558405749f6)

5. I now try to use an `IF ELSEIF STATMENT` like below: But it shows: `RESULT = {"Var1":"DistanceA,20.7"}` even that the Distance is not <=15 or >=30
`Rule1 ON VL53L1X#Distance DO IF (VL53L1X#Distance<=15) var1 DistanceA,%value% ELSE IF (VL53L1X#Distance>=30) var1 DistanceB,%value% ENDIF ENDIF ENDON`
![image](https://github.com/arendst/Tasmota/assets/47510843/d43c0728-45eb-4cb3-8489-60a2a0ed552f)

I am not sure if I am missing anything?

Thank you!

**Comment by sfromis** [comment]
You can **only** use trigger expressions with `#` in the trigger part of a rule, on the left between `on` and `do`. When you have it there, you can then use the value after `do` by using `%value%`.

**Comment by barbudor** [comment]
Step 2 is useless. Why have you done that ?
I told you just to build `tasmota` or `tasmota-4M` with the changes in `user_config_override.h`
You have high risk to break the fine tuned tasmota environment by going to tweak where you are not advised to.

Point 4: Your syntax is wrong `Rule1 ON VL53L1X#Distance DO IF (VL53L1X#Distance<=15) var1 DistanceA,%value% ENDIF ENDON`
`VL53L1X#Distance` is a trigger and the only location where it can go is between `ON` and `DO`
The value of the trigger is `%value%`

The right syntax is `Rule1 ON VL53L1X#Distance DO IF (%value%<=15) var1 DistanceA,%value% ENDIF ENDON`

Of course, this brings you no value comparatively to `Rule1 ON VL53L1X#Distance<=15 DO var1 DistanceA,%value% ENDON` which does the same thing without the need for IF statement

**Reply by barbudor** [reply]
IF statement has no value when it comes to compare the value of the trigger. It has interrest only when you want to include more elements in the comparison

Your Step 5 can also be done without IF statement using 
```
Rule1 ON VL53L1X#Distance<=15 DO var1 DistanceA,%value% ENDON ON VL53L1X#Distance>=30 DO var1 DistanceB,%value% ENDON
```

**Comment by Jason2866** [comment]
The way you tried is not the way it is done correctly with Tasmota. There is no need to change anything in section `[common]`  Your changes will break the setup. Tasmota does not use boards. It uses different settings for flash sizes for the esp8266. For ESP32 just different MCU setups.
Use board https://github.com/arendst/Tasmota/blob/development/platformio_tasmota_env.ini#L23
All can be done in `user_config_override.h` See Tasmota Docu https://tasmota.github.io/docs/Compile-your-build/

**Comment by jackpotcityco** [comment]
> Step 2 is useless. Why have you done that ?
> The way you tried is not the way it is done correctly with Tasmota. There is no need to change anything in section [common] Your changes will break the setup.

That was bad, I shouldn't have changed something there. I was just looking at the file and thought that I needed to specify something there. I will change that back and `re-build` and `flash` this again.

> The right syntax is Rule1 ON VL53L1X#Distance DO IF (%value%<=15) var1 DistanceA,%value% ENDIF ENDON

That worked perfectly now. Also the `IF ELSEIF` statement worked when I changed that using `%value%` and only keep the `"#"` in the trigger.

> Your Step 5 can also be done without IF statement using
> Rule1 ON VL53L1X#Distance<=15 DO var1 DistanceA,%value% ENDON ON VL53L1X#Distance>=30 DO var1 DistanceB,%value% ENDON

Yes, I know this could be done only with `ON DO` as 2 different sub rules. I beleive this was only one example to see that it works to write the `IF ELSEIF` statements as the commands will be different later when I would have actual use of the `IF ELSEIF`

It is amazing to get this `IF ELSEIF ELSE` to work. Wonderful functionality!

Thank you for your help!
